### ISSUE TITLE
Publish sync CJS entrypoint to allow ts-node without --loader

### ISSUE THREAD

**Issue by cspotcode** [google/zx#197]
Is it acceptable for zx to publish a CommonJS entrypoint in addition to the ESM entrypoint?  This would be a big quality-of-life win for anyone combining zx with ts-node, because they would not need to use `--loader` nor to asynchronously load `zx` within their .ts scripts.

The only use of top-level await within `zx` is a call to `which`.  `which` has a sync alternative that can be used to avoid top-level await.

We can still keep both the ESM and CommonJS entrypoints, so consumers can use whichever they prefer.

I'd be happy to send a pull request for this if it will be accepted.

## Expected Behavior

Ability to import zx from CommonJS without await, so that `ts-node ./script.ts` works without needing `--loader`

## Actual Behavior

zx is published as an ESM module, so it must be loaded async.

## Steps to Reproduce the Problem

Create `script.ts` with
```
#!/usr/bin/env ts-node
import {$} from 'zx';
import {without} from 'lodash';
// do stuff
```
`npm install zx @types/lodash lodash`
`npm install -g ts-node typescript`
`ts-node ./script.ts`

## Specifications

- Version: 3.0.0
- Platform: Linux

**Comment by antonmedv** [comment]
So what do you proposing? Change type to cjs in package.json and remove top level await?

**Comment by cspotcode** [comment]
I believe that would work, yes.  When sending the pull request, I would test that using named imports from ESM still works, too.

**Comment by antonmedv** [comment]
I’m actually more agree with Sindre https://github.com/sindresorhus/meta/discussions/15

I prefer tools to support ESM instead of falling back to cjs.

**Comment by cspotcode** [comment]
I understand.  I agree in theory, except that node's ESM support is incomplete, so CommonJS still has more features.  ESM will get there eventually, but it's not there yet.  While node has these feature gaps, it makes ESM on node inconvenient in ways that CJS on node is not.

If anyone reads this in the future and wonders what I mean by "incomplete," I'm referring to the experimental status of `--loader`.  I understand this may confuse people, since node has declared native ECMAScript modules as "stable" even though `--loader` and other ESM-related features are still "experimental."

I published [@cspotcode/zx](https://www.npmjs.com/package/@cspotcode/zx) and shared the code changes in #198

**Comment by rhyek** [comment]
Trying to switch to zx for some scripts in a monorepo where everything else works nicely with ts-node and hit this bump. I have an additional issue where in my script I am importing a library that only provides a cjs entrypoint. I see that zx provides its own `require` which I think might make CJS modules (no specific mention of this in the docs), but if I try to run my script with `zx` it tries to compile everything in my monorepo for some reason (even libraries) and I suspect it's using the wrong (or none) tsconfig.json. Honestly no idea.

Finally, I don't think the nodejs ecosystem is at the point where you can introduce hurdles to people when they are trying to use this library along with things like ts-node and cjs modules. In my case I'll just not use zx, **even though I really want to**. Hopefully you can provide a CJS entrypoint in the future!

This article is interesting: https://redfin.engineering/node-modules-at-war-why-commonjs-and-es-modules-cant-get-along-9617135eeca1

**Comment by cspotcode** [comment]
Does @cspotcode/zx work for you?

**Comment by rhyek** [comment]
Actually, yes. Worked perfectly.

**Comment by antonmedv** [comment]
I didn’t know about exports maps. I think this can be a solution.

**Comment by cspotcode** [comment]
Should we re-open #198?

**Comment by antonmedv** [comment]
Actually, there is another problem: [globby](https://www.npmjs.com/package/globby) is ESM only.

**Comment by antonmedv** [comment]
It looks really ugly to combine CJS and ESM.

**Comment by cspotcode** [comment]
It's not ugly from a consumer's perspective, which is the motivation of #198.  I spent relatively little time implementing zx.  Most of my time is spent consuming it.  So I focus on making consumption easy, knowing the the internal complexity will evaporate in the future when node's ESM support is feature-complete.

**Comment by cspotcode** [comment]
For anyone reading along, async usage of globby remains largely unchanged with #198.

**Comment by antonmedv** [comment]
I need some time to think and research this feature. Let's talk next weekend again. Nope, you understand.

**Comment by cspotcode** [comment]
Of course, sounds good.  And I understand if this change is ultimately rejected; it's not my decision to make.

**Comment by antonmedv** [comment]
Another possibility is to bundle everything into a single .cjs file, and distribute it with zx.

**Comment by cspotcode** [comment]
That's effectively what I did in #198, while preserving the native ESM syntax since it sounded like you wanted to keep using that syntax.

**Comment by cspotcode** [comment]
The .mjs entrypoint makes named exports explicit, so that we avoid relying on node's static analysis: https://nodejs.org/api/esm.html#esm_interoperability_with_commonjs

**Comment by cspotcode** [comment]
> Another possibility is to bundle everything into a single .cjs file, and distribute it with zx.

Oh nevermind, you're suggesting to bundle dependencies such as globby into the cjs file.

**Comment by antonmedv** [comment]
I've created PR https://github.com/google/zx/pull/203

Please, take a look and test. What do you think?
